Add checkpoint saving and loading functionality to training loop

lvermue opened the pull request:
Currently training loops cannot save checkpoints. 
This means that especially in cases of very long training sessions the training cannot be paused and resumed at a later point and that users will lose the entire training progress whenever a crash occurs, e.g. power outage, server reboot or another user trying to use the same GPU.

The solution presented here allows to:
- Automatically store checkpoints during training, by setting the parameter `checkpoint_file`, based on the past time interval since saving last time, which can be adjusted by the parameter `checkpoint_frequency`
- Automatically restore checkpoints from the `checkpoint_file` and resume training where it left off.
- To ensure that the given `checkpoint_file` matches the training loop configuration, a checksum is calculated from the hashed model and optimizer configuration.

To Do:
- [x] Consider logic how the pipeline should interact with the training loop checkpoint functionality
- [x] Write tests

lvermue committed:
Add checkpoint saving and loading functionality to training loop

lvermue committed:
Flake8 and parameters defaults

lvermue committed:
Flake8

cthoyt commented on src/pykeen/training/training_loop.py:
```diff
@@ -673,3 +729,52 @@ def _free_graph_and_cache(self):
+        logger.debug("=> Saving checkpoint.")
+        torch.save(
+            {
+                'epoch': self._epoch,
```
if you try running save_state() and you didn't set a checkpoint file, it will get an AttributeError for `self._epoch`

lvermue replied:
Good catch 👍 That should be fixed now.

cthoyt commented on src/pykeen/training/training_loop.py:
```diff
@@ -673,3 +729,52 @@ def _free_graph_and_cache(self):
+        )
+        logger.info("=> Saved checkpoint.")
+
+    def load_state(self, path: str) -> int:
```
How should this function get used? Would you please provide an example?

lvermue replied:
I have made it a private function now, because imo a user should not fiddle around with these functions. The entire functionality is built to work automatically and consistently by providing a checkpoint file.

lvermue committed:
Refactor internal completed epoch tracking

lvermue committed:
Correct epoch number handling in training loop

lvermue committed:
Add automatic loading of early stoppers from training loop checkpoints

mberr commented on src/pykeen/stoppers/early_stopping.py:
```diff
@@ -191,3 +191,26 @@ def get_summary_dict(self) -> Mapping[str, Any]:
         )
+
+    def _write_from_summary_dict(
+            self,
```
I think we use 4 space continuation indent everywhere else.

mberr commented on src/pykeen/training/training_loop.py:
```diff
@@ -118,6 +125,16 @@ def device(self):  # noqa: D401
+        # https://docs.python.org/3/library/zlib.html
+        return (
+            adler32(str(self.model).encode('utf-8')) & 0xffffffff
+            + adler32(str(self.optimizer).encode('utf-8')) & 0xffffffff
```
I am not sure why we need the bit-wise and. Could you add a comment here?

Also, does `str(model)` capture everything relevant?

lvermue replied:
This is based on the note here https://docs.python.org/3/library/zlib.html#zlib.adler32.
I consider this a quick test, avoiding that some obvious mistakes are made. It returns the used model, loss, regularizer as well as the embedding sizes.
However, the model, loss and regularizer are hashed without their parameters, so this could be added if considered necessary.

cthoyt replied:
Also, why not use something from `hashlib` like the md5 hash? The following code would give a md5 hash string

```python
import hashlib
h = hashlib.md5()
h.update(str(self.model).encode('utf-8'))
h.update(str(self.optimizer).encode('utf-8'))
return h.hexdigest()
```

lvermue replied:
Doesn't really make a difference as long it delivers the same hash each time. But I'm supporting this suggestions, since it allows us to get rid of the bit-magic of zlib.adler32

mberr commented on src/pykeen/training/training_loop.py:
```diff
@@ -179,22 +203,39 @@ def train(
-        )
+        # If a checkpoint file is given we check whether it exists already and load it, if it does
+        if checkpoint_file:
+            if os.path.isfile(checkpoint_file):
```
Likely just a matter of taste, but I prefer the high-level [pathlib](https://docs.python.org/3/library/pathlib.html?highlight=pathlib#module-pathlib) for working with paths. It also adds a type for paths making it clear from type annotations where a path is expected (instead of the generic `str`).

cthoyt replied:
but we don't want to make users import pathlib and instantiate pathlib.Path objects

mberr replied:
sure, but everything internally handled could use them

cthoyt replied:
maybe we could switch that all over in a super PR

lvermue replied:
@mberr 
Do you want to open a feature/enhancement issue?

mberr replied:
#128

mberr commented on src/pykeen/training/training_loop.py:
```diff
@@ -673,3 +742,62 @@ def _free_graph_and_cache(self):
+            stopper_dict = checkpoint['stopper_dict']
+            logger.info(f"=> loaded checkpoint '{path}' stopped after having finished epoch {checkpoint['epoch']}")
+        else:
+            raise FileExistsError(
```
I am not sure whether `FileExistsError` is appropriate here considering its [documentation](https://docs.python.org/3/library/exceptions.html#FileExistsError).

lvermue replied:
Do you have a better suggestion? 😄

mberr replied:
I did not really find anything fitting [here](https://docs.python.org/3/library/exceptions.html). It should be something like an exception raised for corrupt files, or bad input.

Maybe we can define our own exception?

lvermue replied:
I'll look into it.

cthoyt replied:
`ValueError` or `RuntimeError` are fine. If you want, you can subclass one/both of them to give a more specific name

lvermue replied:
I added the CheckpointMismatchError to the training loop.

mberr reviewed (commented):
Generally looks good to me. Most of the comments are no merge-stoppers :slightly_smiling_face:

lvermue committed:
Add training loop checkpoint support for pipelines

lvermue committed:
Fix indentation

cthoyt commented on src/pykeen/training/training_loop.py:
```diff
@@ -179,22 +203,39 @@ def train(
+                logger.info(f"=> no checkpoint found at '{checkpoint_file}'. Creating a new file.")
+
+        # If the stopper loaded from the training loop checkpoint stopped the training, we return those results
+        if getattr(stopper, 'stopped', False):
```
maybe `stopped` should just be set to false in the `NopStopper`

lvermue replied:
I considered that as well. But this solution is more explicit to me, because it shows that only "real" stoppers can have stopped a training loop and thus have the attribute `stopped`.

lvermue committed:
Change training loop checksum creation from adler32 to md5

cthoyt commented on src/pykeen/pipeline.py:
```diff
@@ -699,6 +699,34 @@ def pipeline_from_config(
+    """
+    torch.save(
+        {
+            'random_seed': random_seed,
```
why is the random seed relevant? so it doesn't get overwritten?

lvermue replied:
When we run a pipeline experiment, we save the random_seed as well, since it is important for all random parts. When the user starts a pipeline without providing a random seed, a random_seed is generated say `random_seed=1`. When the users pauses the training and resumes it later on by starting the same pipeline again, a new random seed would be generated, say `random_seed=2`. When the experiment finishes, it would state that the experiment was run with `random_seed=2`. This would not only be incorrect, but also not allow the user to reproduce the experiment.
This problem is remedied by storing and re-using the random seed from the pipeline helper file.

mberr replied:
There is still a difference in starting a run with a fixed random seed and letting it run until the end, and stopping in between and restarting from a checkpoint: When restarting from a checkpoint, the last random state does not get restored. Thus, all random parts, e.g. negative sampling, will yield different results than they had done when running through.

I am not sure whether we can really solve this problem though, or if it would be better to guarantee it for running an experiments, but not when restoring from checkpoints.

cthoyt replied:
it might be possible to acquire the current random state from PyTorch... but like once you're at this point, you're probably not really worried about **re**producibility so much as reproducibility for the first time lol

cthoyt replied:
So the `pytorch.Generator()` object could address this problem, but we'd have to do a pretty major refactoring to use our own instance of one of those.

Ref: https://github.com/pytorch/pytorch/issues/17079

lvermue replied:
This should be fixed with https://github.com/pykeen/pykeen/pull/123/commits/fc9a5a9fb01e6a9edd5db92dbc498b0f27fe8b17

cthoyt replied:
That's awesome!!

cthoyt commented on src/pykeen/training/training_loop.py:
```diff
@@ -412,8 +464,23 @@ def _train(  # noqa: C901
+                # If a checkpoint file is given, we check whether it is time to save a checkpoint
+                if checkpoint_file:
+                    minutes_since_last_checkpoint = (time.time() - last_checkpoint) // 60
+                    if minutes_since_last_checkpoint >= checkpoint_frequency:
```
there's some duplication of code here, can it be reduced?

lvermue replied:
I considered creating a function, but that function would need the `checkpoint_file`, `last_checkpoint` and `stopper` as argument. In addition, it would have to return the variable `last_checkpoint`, which would have to be assigned within the training_loop. Imo this would be a worse solution than it is right now.

lvermue replied:
The only other possibility I came up with is
```
            should_stop = False
            if stopper is not None and stopper.should_evaluate(epoch) and stopper.should_stop(epoch):
                should_stop = True

            # If a checkpoint file is given, we check whether it is time to save a checkpoint
            if checkpoint_file:
                minutes_since_last_checkpoint = (time.time() - last_checkpoint) // 60
                if minutes_since_last_checkpoint >= checkpoint_frequency:
                    self._save_state(path=checkpoint_file, stopper=stopper)
                    last_checkpoint = time.time()
            
            if should_stop:
                return self.losses_per_epochs
```

lvermue replied:
Which one do you prefer?

lvermue replied:
I have added it now with https://github.com/pykeen/pykeen/pull/123/commits/7b84f70586c7e80f5277ffa8e26304da81f01825.

lvermue committed:
Fix missing random_seed failure, if no training config file is provided

lvermue committed:
Flake8, refactor code and add CheckpointMismatchError

lvermue committed:
Fix flake8

cthoyt commented on src/pykeen/pipeline.py:
```diff
@@ -699,6 +699,34 @@ def pipeline_from_config(
+    )
+
+
+def load_pipeline_checkpoint_helper_file(path: str) -> Mapping[str, Any]:
```
does this really need its own function?

lvermue replied:
I agree that it's a bit overkill right now, but it was build with expandability in mind for the entire pipeline and it also adds some clarity to the code about what these files actually contain.
But for now I would also be finde with only making the straight function calls without the helper functions.

cthoyt replied:
I think there's an old rule for software dev not to implement new stuff until you really need it. Another way we could go about this is to use TypedDicts / type annotations to add into the code directly where torch.load is called

cthoyt committed:
Update exception

cthoyt committed:
Add outline for checkpoint tutorial

lvermue committed:
Correct random state recovery

lvermue committed:
Merge branch 'allow_training_checkpoints' of https://github.com/pykeen/pykeen into allow_training_checkpoints

cthoyt commented on src/pykeen/pipeline.py:
```diff
@@ -823,9 +851,26 @@ def pipeline(  # noqa: C901
+    # containing the used random_seed to ensure reproducible results
+    if training_kwargs and training_kwargs.get('checkpoint_file'):
+        checkpoint_file = training_kwargs.get('checkpoint_file')
+        pipeline_checkpoint_helper_file = f"{checkpoint_file}_pipeline_helper_file"
```
let's make a better naming convention for this

lvermue replied:
Any ideas?

cthoyt commented on src/pykeen/pipeline.py:
```diff
@@ -823,9 +851,26 @@ def pipeline(  # noqa: C901
+        pipeline_checkpoint_helper_file = f"{checkpoint_file}_pipeline_helper_file"
+        if os.path.isfile(pipeline_checkpoint_helper_file):
+            pipeline_checkpoint_helper_dict = load_pipeline_checkpoint_helper_file(pipeline_checkpoint_helper_file)
+            random_seed = pipeline_checkpoint_helper_dict['random_seed']
```
since the other function sets the states directly, is saving it in a file even necessary now?

lvermue replied:
It would for the least still be relevant to have the pipeline to correctly return the used random seed.
Aside that, it would give us the guarantee that whatever random method is involved before we resume the training loop also uses the correct random seed.

lvermue commented on src/pykeen/training/training_loop.py:
```diff
@@ -673,3 +745,64 @@ def _free_graph_and_cache(self):
+                'stopper_dict': stopper_dict,
+                'random_state': random.getstate(),
+                'np_random_state': np.random.get_state(),
+                'torch_random_state': torch.random.get_rng_state(),
```
@cthoyt @mberr @mali-git 
Considering everything relevant here - Don't we also have stateful regularizers?
And is there anything you guys can think of?

mberr replied:
Yes, there are e.g. the weights of the regularizer, cf. 
https://github.com/pykeen/pykeen/blob/dc247da229bf0788b526b151c98148e0e617d64e/src/pykeen/regularizers.py#L27-L85

mberr replied:
But, since the regularizer is a module, and its an attribute of `Model`, its pytorch weights should get saved. From [here](https://discuss.pytorch.org/t/what-is-the-difference-between-register-buffer-and-register-parameter-of-nn-module/32723) it seems that this also holds for buffers.

lvermue committed:
Remove pipeline checkpoint helper file help function

lvermue committed:
Remove torch save helper file

lvermue committed:
Fix function argument for torch save

lvermue committed:
Add units tests for training loop checkpoint

cthoyt committed:
Merge branch 'master' into allow_training_checkpoints

lvermue committed:
Add unit tests for checkpoints

lvermue committed:
Fix flake8

lvermue committed:
Add failure fallback checkpoints to training loop

mberr commented on src/pykeen/training/training_loop.py:
```diff
@@ -4,7 +4,7 @@
 import gc
 import logging
-import os
+import pathlib
```
:heart:

lvermue committed:
Correct checkpoint root dir handling

lvermue committed:
Add implicit random seed handling from checkpoints in pipeline

cthoyt committed:
Code cleanup

Store results of get() to make code cleaner

cthoyt committed:
More refactoring

cthoyt commented on src/pykeen/training/training_loop.py:
```diff
@@ -673,3 +797,67 @@ def _free_graph_and_cache(self):
+                'random_state': random.getstate(),
+                'np_random_state': np.random.get_state(),
+                'torch_random_state': torch.random.get_rng_state(),
+                'torch_cuda_random_state': torch.cuda.get_rng_state(),
```
This fails if torch wasn't compiled with CUDA - this means that this will fail on machines using CPU

mberr replied:
There is also `torch.cuda.is_available()` which could be used.

lvermue committed:
Add CPU/GPU random state differentiation

lvermue committed:
Merge branch 'allow_training_checkpoints' of https://github.com/pykeen/pykeen into allow_training_checkpoints

cthoyt committed:
Workaround for CUDA rng state

cthoyt committed:
Refactor tests

cthoyt committed:
Remove unnecessary stuff

cthoyt committed:
Unnest logic

cthoyt commented on tests/test_training.py:
```diff
@@ -86,6 +87,9 @@ class TrainingLoopTests(unittest.TestCase):
         """Instantiate triples factory and model."""
         self.triples_factory = Nations().training
+        self.random_seed = 123
+        self.checkpoint_file = f"PyKEEN_training_loop_test_checkpoint_{da}.pt"
```
da?

lvermue replied:
Not entirely finished apparently

cthoyt commented on src/pykeen/training/training_loop.py:
```diff
@@ -179,22 +213,62 @@ def train(
+        if checkpoint_file:
+            checkpoint_file_path = checkpoint_root.joinpath(checkpoint_file)
+            if checkpoint_file_path.is_file():
+                stopper_dict = self._load_state(path=checkpoint_file_path)
```
this is a bit confusing since _load_state() has side effects and a return value

cthoyt committed:
Improve typing and safety

Since the type of `stopper` is Optional[Stopper], it doesn't know that this is a function of the EarlyStopper subclass. I added the function to the superclass that does nothing, so it will now make sense for the type checker.

Also, I checked for none in case of misconfiguration

lvermue committed:
Fix testing

lvermue committed:
Fix pipeline checkpoint unit tests

lvermue committed:
Merge branch 'master' into allow_training_checkpoints

lvermue committed:
Fix usage of forbidden characters for Windows in filepaths

lvermue committed:
Refactor loading of states for the training loop and stoppers

lvermue committed:
Fix flake8

lvermue committed:
Fix handling of stopper state dictionaries

lvermue committed:
Merge branch 'master' into allow_training_checkpoints

lvermue committed:
Fix flake8

cthoyt commented on docs/source/tutorial/checkpoints.rst:
```diff
@@ -0,0 +1,9 @@
+Using Checkpoints
+=================
+Why does someone want to use checkpoints?
```
@lvermue write up this tutorial please

cthoyt commented on src/pykeen/constants.py:
```diff
@@ -3,9 +3,15 @@
 ]
 
 PYKEEN_HOME = os.environ.get('PYKEEN_HOME') or os.path.join(os.path.expanduser('~'), '.pykeen')
+PYKEEN_DEFAULT_CHECKPOINT = "PyKEEN_just_saved_my_day.pt"
```
perfect name

cthoyt commented on src/pykeen/training/training_loop.py:
```diff
@@ -679,3 +810,85 @@ def _free_graph_and_cache(self):
+                'model_state_dict': self.model.state_dict(),
+                'optimizer_state_dict': self.optimizer.state_dict(),
+                'checksum': self.checksum,
+                'random_seed': self.model._random_seed,
```
is this necessary for continuing training? Since the current state when it's saved is also here?

cthoyt replied:
in general we don't do a perfect job at storing all configuration settings, but I don't know if having the training loop take care of this makes so much sense

lvermue replied:
> is this necessary for continuing training? Since the current state when it's saved is also here?

As discussed before, the `random_seed` is required for documentation purposes.

lvermue replied:
> in general we don't do a perfect job at storing all configuration settings, but I don't know if having the training loop take care of this makes so much sense

If we want, we could take over all "training loop relevant" parts from the pipeline and log them in the training loop as well.

cthoyt commented on tests/test_pipeline.py:
```diff
@@ -108,6 +109,74 @@ def test_predict_all_with_novelties(self):
+        date_string = str(datetime.now()).replace('.', '_').replace(':', '_')
+        self.checkpoint_file = f"PyKEEN_training_loop_test_checkpoint_{date_string}.pt"
+
+    def test_pipeline_lcwa_resumption(self):
```
these tests share a lot of code, can you make it into a helper function?

cthoyt commented on tests/test_training.py:
```diff
@@ -86,6 +88,10 @@ class TrainingLoopTests(unittest.TestCase):
         self.triples_factory = Nations().training
+        self.random_seed = 123
+        date_string = str(datetime.now()).replace('.', '_').replace(':', '_')
+        self.checkpoint_file = f"PyKEEN_training_loop_test_checkpoint_{date_string}.pt"
```
might  be better to use in combination with a temporary directory during the tests

mberr replied:
https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory

cthoyt commented on src/pykeen/training/training_loop.py:
```diff
@@ -341,84 +438,118 @@ def _train(  # noqa: C901
+
         # Training Loop
         for epoch in epochs:
-            # Enforce training mode
```
@lvermue thanks for taking care of merging the updates on master back into this, I would have done it for you but I didn't want to mess any of the new code up

cthoyt commented on src/pykeen/training/training_loop.py:
```diff
@@ -181,22 +215,69 @@ def train(
+
+        # If a checkpoint file is given, it must be loaded if it exists already
+        save_checkpoints = False
+        if checkpoint_file:
```
There's an awful lot of branching logic here. are there unit tests for all of them?

lvermue replied:
Imo unit tests wouldn't make sense here, as all this branching logic simply is to inform the user about what is happening when using checkpoints combined with the consequences of different choices or possible misconfigurations. Or do you have something else in mind?

cthoyt replied:
I think it's okay for now. After a few rounds of refactoring, it's much more obvious

mberr commented on tests/test_training.py:
```diff
@@ -121,3 +127,98 @@ def test_blacklist_loss_on_slcwa(self):
         with self.assertRaises(TrainingApproachLossMismatchError):
             NaNTrainingLoop(model=model, patience=2)
+
+    # Add docu
```
This looks like a fixed todo

mberr reviewed (commented):
I do not have additional review comments on top of those @cthoyt already mentioned.

lvermue committed:
Merge branch 'master' into allow_training_checkpoints

lvermue committed:
Fix unit tests

lvermue committed:
Refactor pipeline unit tests

lvermue committed:
Refactor training loop unit tests

lvermue committed:
Fix flake8

lvermue committed:
Merge branch 'master' into allow_training_checkpoints

lvermue committed:
Add saving of checkpoints after successful training

lvermue committed:
Add usage of temporary directories for unit tests

lvermue committed:
Add checkpoint documentation and correct failure checkpoint handling

PyKEEN-bot committed:
Trigger CI

cthoyt committed:
Add missing variable default value

cthoyt committed:
Get rid of tqdms

cthoyt committed:
Pass flake8

cthoyt committed:
Use class teardown for handling temporary directory

Trigger CI

cthoyt committed:
Update docs

cthoyt committed:
Update argument names and type hints

there are a lot of places where either a string or a pathlib.Path is acceptable

cthoyt commented on src/pykeen/training/training_loop.py:
```diff
@@ -189,22 +229,76 @@ def train(
+        # If a checkpoint file is given, it must be loaded if it exists already
+        save_checkpoints = False
+        checkpoint_path = None
+        if checkpoint_name:
```
can there be a reasonable default for checkpoint_name, then like in the helper method, just pass a flag `save_checkpoints=True` so we don't have to use this logic?

lvermue replied:
Imo it is cleaner to follow the logic that explicitly giving a `checkpoint_name` indicates that someone wants to save/reload that specific checkpoint.

cthoyt replied:
~Then why not just have the user pass the fully formed path instead of both the directory and the name?~

Ehh I get that this gives a nice level of granularity, I'm just advocating for the lazy user (aka @cthoyt)

cthoyt commented on docs/source/tutorial/checkpoints.rst:
```diff
@@ -0,0 +1,147 @@
+Again, assuming that e.g. this pipeline crashes after 200 epochs, you can simply execute **the same code** and the
+pipeline will load the last state from the checkpoint file and continue training as if nothing happened.
+
+.. todo:: Tutorial on recovery from hpo_pipeline.
```
Maybe for a later PR?

lvermue replied:
Yes. Basically the hpo_pipeline supports saving checkpoints through the pipeline, but that is specific to the training loop itself. Supporting to resume a cancelled hpo_pipeline would be an entirely different story.

cthoyt replied:
I wonder if this `.. todo` directive breaks the build

cthoyt commented on src/pykeen/training/training_loop.py:
```diff
@@ -189,22 +229,76 @@ def train(
+        checkpoint_on_failure_file_path = None
+        if checkpoint_on_failure:
+            # In case a checkpoint frequency was set, we warn that no checkpoints will be saved
+            # TODO there's a method for this: datetime.now().strftime('your fmt string')
```
this line can be greatly improved with builtin functionality

cthoyt commented on src/pykeen/training/training_loop.py:
```diff
@@ -265,9 +363,17 @@ def _train(  # noqa: C901
+        :param checkpoint_frequency:
+            The frequency of saving checkpoints in minutes. Setting it to 0 will save a checkpoint after every epoch.
+        :param checkpoint_on_failure_file_path:
+            TODO
```
needs docs

cthoyt replied:
also, it used to say moin. so Moin right back at ya

lvermue replied:
https://en.wikipedia.org/wiki/Moin

cthoyt replied:
I know moin ;)

cthoyt commented on tests/test_pipeline.py:
```diff
@@ -108,6 +109,76 @@ def test_predict_all_with_novelties(self):
+        """Test whether the resumed pipeline creates the same results as the one shot pipeline."""
+        checkpoint_directory = self.temporary_directory.name
+
+        # TODO check that there aren't any checkpoint files already existing in the right place
```
can we add some more tests here that illustate what's going on between each step? It's pretty high level and won't be helpful to someone without them having to fully grok the code

lvermue replied:
Wouldn't that be the tests in *test_training.py*?

lvermue replied:
For the same reason as given in https://github.com/pykeen/pykeen/pull/123#discussion_r534073661 , these tests only exist to see whether the pipeline handles those training loop checkpoints properly. The main functionality fully resides in the training loop.

cthoyt replied:
okay we don't have to make this part overcomplicated. Maybe a code comment saying to check the other test then?

PyKEEN-bot committed:
Trigger CI

lvermue committed:
Add datetime formatting

cthoyt commented on docs/source/tutorial/checkpoints.rst:
```diff
@@ -0,0 +1,147 @@
+To show how checkpoints are used with PyKEEN let's look at a simple example of how a model is setup.
+For fixing possible errors and safety fallbacks please also look at :ref:`word_of_caution`.
+
+.. code-block:: python
```
wouldn't these tutorials be more useful for users if they started by being centered on the `pipeline` and then at the end gave some insight into the underlying implementation?

lvermue replied:
The reason it is kept right now is due to the fact that the checkpoint functionality is only a *true* training loop functionality, because even though the pipeline supports using training loop checkpoints, it is not a *true* pipeline checkpoint.

cthoyt replied:
That's true, but I don't think the beginning of a tutorial section of the documentation benefits from being pedagogical. The technical part can be in the reference, or at the end of the tutorial to help users who understand how to use the simple parts and want to understand how it works. I think one place this worked really well was the [First Steps](https://pykeen.readthedocs.io/en/latest/tutorial/first_steps.html) tutorial, which ended with the [Beyond the Pipeline](https://pykeen.readthedocs.io/en/latest/tutorial/first_steps.html#beyond-the-pipeline) section.

Could you elaborate on the difference you mean between a training loop checkpoint vs a pipeline checkpoint?

cthoyt commented on tests/test_pipeline.py:
```diff
@@ -108,6 +109,76 @@ def test_predict_all_with_novelties(self):
+        self.checkpoint_name = "PyKEEN_training_loop_test_checkpoint.pt"
+        self.temporary_directory = tempfile.TemporaryDirectory()
+
+    def tearDown(self) -> None:
```
@lvermue note I changed the temporary directory usage here to use the test case's tearDown functionality

lvermue replied:
Why that? 👀 
Now there will be orphans when the unit tests are cancelled.

cthoyt replied:
no it's smart and runs teardown even if they're cancelled

lvermue committed:
Add docs for checkpoint_on_failure_file_path

cthoyt committed:
Merge branch 'master' into allow_training_checkpoints

cthoyt committed:
Update constants

lvermue committed:
Change temp dir creation and teardown during unit tests

lvermue committed:
Merge branch 'allow_training_checkpoints' of https://github.com/pykeen/pykeen into allow_training_checkpoints

cthoyt commented on tests/test_training.py:
```diff
@@ -183,7 +186,7 @@ def _test_checkpoints(self, training_loop_type: str, checkpoint_directory: str):
             batch_size=self.batch_size,
             checkpoint_name=self.checkpoint_file,
-            checkpoint_directory=checkpoint_directory,
+            checkpoint_directory=self.temporary_directory,
```
if you use TemporaryDirectory() as a context manager, the name is returned as the context. However if you save as a variable, you need `self.temporary_directory.name`

lvermue committed:
Update the checkpoint tutorial

PyKEEN-bot committed:
Trigger CI

lvermue committed:
Fix temp dir name handling

PyKEEN-bot committed:
Trigger CI

cthoyt commented on docs/source/tutorial/checkpoints.rst:
```diff
@@ -2,15 +2,159 @@ Using Checkpoints
+        training_kwargs=dict(num_epochs=1000),
+    )
+
+To enable checkpoints all you have to do is adding a ``checkpoint_file`` argument to the ``training_kwargs``.
```
should be `checkpoint_name`

cthoyt replied:
Resolved in ed6b98e3

cthoyt committed:
Small fixes in docs

Trigger CI

cthoyt reviewed (approved):
Ready to merge once CI passes. Thanks @lvermue for writing an excellent tutorial!

lvermue committed:
Merge branch 'master' into allow_training_checkpoints

PyKEEN-bot committed:
Trigger CI
